odoo/models.py

```diff
@@ -5521,6 +5521,33 @@ def onchange(self, values, field_name, field_onchange):
 
         return result
 
+    #
+    # Cleanup
+    #
+
+    @api.model
+    def _expunge_bad_records(self):
+        """
+        Expunge bad records from the current model.
+
+        This method, when overridden, is used to remove records of models
+        which may have database constraints and could not be removed normally
+        through the unlink method.
+
+        E.g.: Model A and Model B both override the unlink method, Model B
+              inherits from Model A and thus in its unlink override, it calls
+              Model A's unlink. If Model B's unlink fails (i.e. raises an 
+              exception) before calling Model A's unlink, there could be
+              some records that still reference Model B left behind, even
+              though Model B has been removed, and those records
+              could be troublesome.
```

Comment:
The example is not clear to me :stuck_out_tongue:
The first sentence refers to inheritance, where records in model A should be gone because of corresponding records in model B are gone. The second second sentence suggest model A has generic references like `res_model`, `res_id` fields that have no database foreign key support...